### PR TITLE
feat!: add packet forward middleware

### LINKED ISSUE #2831: Adopt the packet forward middleware
A common feature request for teams building on IBC is for the PFM to be added to the middleware of celestia-app

### PR DESCRIPTION
**Pull request by ninabarbakadze**
closes #2831 
opens #3241

### CODE REVIEW

**Comment by rootulp** [comment]
> closes - https://github.com/celestiaorg/celestia-app/issues/2831
> opens - https://github.com/celestiaorg/celestia-app/issues/3241

I think the hyphen in the middle is going to prevent #2831 from actually closing when this PR merges. I'd suggest moving the hyphens.

Ref: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: app/app.go [hunk 1]
```diff
@@ -201,6 +206,8 @@ type App struct {
 	upgradeHeight int64
 	// used to define what messages are accepted for a given app version
 	MsgGateKeeper *ante.MsgVersioningGateKeeper
+
+	PacketForwardKeeper *packetforwardkeeper.Keeper
```
Comment (rootulp) [inline app/app.go:210]: [follow-up][optional] the fields in this struct don't follow a clear order. We could move these two keeprs into the group below the line:
```
// keepers
```

but it's an extreme nit so nvm

File: specs/src/specs/state_machine_modules.md [hunk 2]
```diff
@@ -24,8 +24,9 @@ Celestia app is built using the cosmos-sdk, and follows standard cosmos-sdk modu
 - [feegrant](https://github.com/celestiaorg/cosmos-sdk/blob/v1.14.0-sdk-v0.46.11/x/feegrant/spec/README.md)
 - [genutil](https://github.com/celestiaorg/cosmos-sdk/tree/v1.14.0-sdk-v0.46.11/x/genutil) (no spec)
 - [gov](https://github.com/celestiaorg/cosmos-sdk/blob/v1.14.0-sdk-v0.46.11/x/gov/spec/README.md)
-- [params](https://github.com/celestiaorg/cosmos-sdk/blob/v1.14.0-sdk-v0.46.11/x/params/spec/README.md)
 - [ibc](https://github.com/cosmos/ibc/blob/f990a7f96eb7753c2fabbd49ed50b64d3a807629/README.md)
+- [pacletforward](https://github.com/cosmos/ibc-apps/blob/main/middleware/packet-forward-middleware/README.md)
```
[the remark this answers is not in the document]
Comment (rootulp) [reply specs/src/specs/state_machine_modules.md:28]: [nit] there's a typo here. Also is the module name `packetforward` or `packetforwardmiddleware`? I think the latter based on [this line](https://github.com/cosmos/ibc-apps/blob/90947a1303b6d1ecb003efba1bbc62bd8de3df37/middleware/packet-forward-middleware/packetforward/types/keys.go#L7) so

```suggestion
- [packetforwardmiddleware](https://github.com/cosmos/ibc-apps/blob/main/middleware/packet-forward-middleware/README.md)
```

File: test/packetforward/packetforward_upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	version "github.com/tendermint/tendermint/proto/tendermint/version"
+	dbm "github.com/tendermint/tm-db"
+)
+
+// TestPacketForwardMiddlewareAgainstAppUpgrades verifies that the PFM module's params are overridden during an
+// upgrade from v1 -> v2 and correctly set when upgrading to v2.
```
Comment (rootulp) [inline test/packetforward/packetforward_upgrade_test.go:26]: [nit] the second half is redundant

```suggestion
// TestPacketForwardMiddlewareAgainstAppUpgrades verifies that the PFM module's params are overridden during an
// upgrade from v1 -> v2.
```

**Review by rootulp — APPROVED** [review]
Overall LGTM! Great work

**Review by evan-forbes — APPROVED** [review]
:+1: LGTM

**Review by rootulp — APPROVED** [review]
Still LGTM

**Pull request merged**